jaeger_client/metrics.py

```diff
@@ -21,69 +21,155 @@
 from __future__ import absolute_import
 
 
-class Metrics(object):
-    """Provides an abstraction of metrics reporting framework."""
+class MetricsFactory(object):
+    """Generates new metrics."""
 
-    # TODO do we want to automatically include host name?
-    prefix = 'jaeger'
+    def __init__(self, count=None, gauge=None, timing=None, tags=None):
+        """
+        :param count: function (key, value) to emit counters
+        :param gauge: function (key, value) to emit gauges
+        :param timing: function (key, value) to emit timings
+        :param tags: {k:v} dictionary
+        """
+        self._count = count
+        self._gauge = gauge
+        self._timing = timing
+        self._tags = tags
+        if not callable(self._count):
+            self._count = _noop
+        if not callable(self._gauge):
+            self._gauge = _noop
+        if not callable(self._timing):
+            self._timing = _noop
 
-    TRACES_STARTED_SAMPLED = '%s.traces-started.sampled' % prefix
-    """Number of traces started by this tracer as sampled"""
+    def counter(self, name, tags=None):
+        """
+        Generates a new counter from the given name and tags.
+        :param name: name of the counter
+        :param tags: tags for the counter
+        :return: a callable function which takes the value to increase
+        the counter by ie. def increment(value)
+        """
+        raise NotImplementedError
 
-    TRACES_STARTED_NOT_SAMPLED = '%s.traces-started.not-sampled' % prefix
-    """Number of traces started by this tracer as not sampled"""
+    def timer(self, name, tags=None):
+        """
+        Generates a new timer from the given name and tags.
+        :param name: name of the timer
+        :param tags: tags for the timer
+        :return: a callable function which takes the duration to
+        record ie. def record(duration)
+        """
+        raise NotImplementedError
 
-    TRACES_JOINED_SAMPLED = '%s.traces-joined.sampled' % prefix
-    """Number of externally started sampled traces this tracer joined"""
+    def gauge(self, name, tags=None):
+        """
+        Generates a new gauge from the given name and tags.
+        :param name: name of the gauge
+        :param tags: tags for the gauge
+        :return: a callable function which takes the value to update
+        the gauge with ie. def update(value)
+        """
+        raise NotImplementedError
+
+    def _merge_tags(self, tags=None):
+        if not self._tags:
+            return tags
+        tags_cpy = self._tags.copy()
+        tags_cpy.update(tags)
+        return tags_cpy
+
+
+class NoopMetricsFactory(MetricsFactory):
```

Comment:
I don't understand why you need this class. If you change the base class to actually invoke the functions it gets as constructor arguments from the main methods (instead of raising not implemented), then the base class becomes Noop (especially since you default to _noop it it already).